Pr/0001 link list in at cmd notif

pkchan opened the pull request:
The original at_cmd module supports only one record notification handler, whereas the design in this PR is a link-list which can support infinite number of handlers theoretically.

The purpose is to allow for numerous modules to receive AT-command notifications at the same time.

A new interface at_cmd_remove_notification_handler() has also been created to remove an entry from the list. The only module which needs to have notification handler removed is lte_lc. It is also changed accordingly.

rlubos commented:
@lemrey @evenl @metc You may find this interesting for the LWM2M library.

rlubos commented on drivers/at_cmd/at_cmd.c:
```diff
@@ -48,6 +54,133 @@ struct callback_work_item {
+	}
+
+	/* Go through each element in the list. */
+	k_sem_take(&cmd_pending, K_FOREVER);
```
Reusing the `cmd_pending` semaphore for protecting the linked list might result in a deadlock (in case the command was waiting for a response while notification arrived). We should use a different semaphore (or mutex preferably) to protect the list. At this point, commands and notifications should not interfere with each other, so it should be safe to have a separate synchronization primitive for notifications.

pkchan replied:
The adding and removing of nodes to the master list (handler_list) are now divided into two phases:
1. Queue to temporary lists handler_to_add and handler_to_remove. It can add a node to handler_to_add, so that the node will be added to handler_list, or it can add a node to handler_to_remove, so that the corresponding node in handler_list will be removed.
2. When the handlers from handler_list are not running or have finished running, the handler_list is then maintained according to the nodes stored in handler_to_add and handler_to_remove. This is done in function maintain_list(). Semaphore list_pending has been created for such purpose.

This way, deadlock does not happen when you try to call at_cmd_set/remove_notification_handler() from the handlers themselves.

rlubos replied:
While I understand your motivation, I can't shake the feeling that we're highly complicating the code for a little benefit. 

Sure, giving extra flexibility for using notification_handler API is nice, but I'm not aware of an existing use case for this (@lemrey @evenl perhaps you know about something?). But now, instead of a single list, we end up in a quite complex design, with three lists and non-trivial to understand (and protect) synchronization mechanism. 

I'm really not in favour of introducing such complexity, especially that all I've asked for was a separate mutext/semaphore to protect list operations. But that's only my opinion, let's hear from other reviewers.

rlubos commented on drivers/at_cmd/at_cmd.c:
```diff
@@ -48,6 +54,133 @@ struct callback_work_item {
+}
+
+/**@brief Dispatcher for notification handlers. */
+static void notification_handler (char *response)
```
Excess space before `(`

pkchan replied:
Removed.

rlubos replied:
Still present

rlubos commented on drivers/at_cmd/at_cmd.c:
```diff
@@ -48,6 +54,133 @@ struct callback_work_item {
+/**@brief Find notification handler. */
+static notif_handler_t *find_notif_handler(at_cmd_handler_t handler)
+{
+	notif_handler_t *l_notif_handlers = notif_handlers;
```
Before going further,

I'm not sure if you are aware, but Zephyr provides a single-linked list implementation:
https://github.com/zephyrproject-rtos/zephyr/blob/master/include/sys/slist.h

It provides quite a few convenience functions and is already used in a few places in NCS - any chance to use it here as well?

Just note that the API is not thread-safe, so synchronization primitives would be needed anyway.

pkchan replied:
Used slist primitives, though it takes longer in the development process to double-check the behaviors of such implementation.

As mentioned [here](https://github.com/NordicPlayground/fw-nrfconnect-nrf/pull/1163#discussion_r326614614), Semaphore has been used to protect slist.

jtguggedal commented:
General note: supporting multiple notification handlers is a good and needed change. Possibly beyond the scope of this PR is that each user of the `at_cmd` module should only receive notifications that it has enabled. As I see it, this change will send all notifications to all users, regardless of if the user has enabled it or not. I think we should aim for absolute separation between users, emulating that each user has its own socket.

rlubos commented on drivers/at_cmd/at_cmd.c:
```diff
@@ -289,12 +416,28 @@ int at_cmd_write(const char *const cmd,
+	k_sem_give(&cmd_pending);
 
-	notification_handler = handler;
+	if (rc == 0 && k_sem_count_get(&list_pending) != 0) {
```
I'm afraid that `k_sem_count_get(&list_pending) != 0` check is not enough for assuring thread-safety. In case two or more threads use `at_cmd_set_notification_handler`/`at_cmd_remove_notification_handler` APIs and one of them is preempted while inside `maintain_list()`, there's a risk of violating one of the lists (if another thread calls these APIs).

To assure thread-safety, you need to actually take the semaphore, otherwise, we're open for (very rare and hard to reproduce) failures (very often fatal).

pkchan replied:
Up to now, I do not see any direction on how to make it "good", since there is no thread-safe link-list implementation at hand.

To go ahead, I see several possibilities:
1. Leave it as is, and keep the change un-merged.
2. Restrict the use case so that only appending of the list is allowed.
3. at_cmd_set_notification_handler/at_cmd_remove_notification_handler simply return -EBUSY when the dispatcher is being run. Up to now, these function calls only happen during initialization phase. It doesn't make sense to have them called during run-time.
4. Ignore mutual exclusion. As mentioned before, there is no use-case to add elements during runtime.

rlubos commented on drivers/at_cmd/at_cmd.c:
```diff
@@ -26,13 +27,22 @@ static K_THREAD_STACK_DEFINE(socket_thread_stack, \
 
 static struct k_thread  socket_thread;
-static at_cmd_handler_t notification_handler;
+static sys_slist_t      handler_list;
```
I'd call it `notification_handler_list` to explicitly differentiate it from command handler.

pkchan commented:
Here is a summary of issues raised by @rlubos :
1. use a dedicated mutex to protect the list (ref: [here](https://github.com/NordicPlayground/fw-nrfconnect-nrf/pull/1163#discussion_r326191308)); Don't use semaphores (ref: [here](https://github.com/NordicPlayground/fw-nrfconnect-nrf/pull/1163#discussion_r326631541));
2. simplify the design (ref: [here](https://github.com/NordicPlayground/fw-nrfconnect-nrf/pull/1163#discussion_r326644001));
3. Remove excess space before ( (ref: [here](https://github.com/NordicPlayground/fw-nrfconnect-nrf/pull/1163#discussion_r326192747));
4. Use slist (ref: [here](https://github.com/NordicPlayground/fw-nrfconnect-nrf/pull/1163#discussion_r326196792));
5. Remove prefixes like "l_" (ref: [here](https://github.com/NordicPlayground/fw-nrfconnect-nrf/pull/1163#discussion_r326623388));
6. Use notification_handler_list as the list name (ref: [here](https://github.com/NordicPlayground/fw-nrfconnect-nrf/pull/1163#discussion_r326653145)).

**Proposal:**
- Only one list : notif_handler_list;
- Non-existing handlers can be added to the list; (Use sys_slist_append());
- Existing handlers can be removed from the list; (Use sys_slist_remove());
- Use mutex to protect both sys_slist_append() and sys_slist_remove();
- Reading of the list need not be protected.

I doubt if list reading need to be protected or not. In case where a node is removed, the memory is freed and can be reclaimed by someone else. The node pointer can become an invalid pointer suddenly, which can be dangerous.

rlubos commented:
@pkchan Thank you for this summary, it seems to cover my overall feedback well. Regarding the:

> Reading of the list need not be protected.

I think that you've already concluded, that we **do need** to protect the loop during the readout, otherwise it can be broken by other thread. There are two mechanisms though that will help us implement that w/o making the implementation too complex:
 
1. https://docs.zephyrproject.org/latest/reference/kernel/synchronization/mutexes.html#reentrant-locking - if we protect the list readout with a mutex, this feature will allow us to call the set/remove notification API (protected with the same mutex) from the notification_handler thread, while still protecting the access from other thread (hence protecting the list pointers from being damaged).

2. https://github.com/zephyrproject-rtos/zephyr/blob/8f1780f6dee29dc2b2db01744e328a1a4a20d0b8/include/sys/slist.h#L160 - this loop helper will allow us to safely operate on the currently called callback node - the callback handler should be able to safely remove itself from the list (IMO that's the most reasonable use-case), w/o breaking the loop.

With above, the only restriction would be that the callback handler is not allowed to remove callbacks other than itself (which could break the current for_each loop). Otherwise, all operations should be safe (unless I miss something).

Please let me know if you find a flaw in the above thought, or think that it's wrong at some point.

pkchan commented:
I agree to use both recursive mutex and SYS_SLIST_FOR_EACH_CONTAINER_SAFE to protect the list when reading. In the comments of at_cmd_remove_notification_handler(), there will be some sentences like "Restrictions: When handler is running, one can only remove the handler itself, but not other handlers."

Then the proposal becomes:

- Only one list : notif_handler_list;
- Non-existing handlers can be added to the list; (Use sys_slist_append());
- Existing handlers can be removed from the list; (Use sys_slist_remove());
- **Use mutex to protect sys_slist_append(), sys_slist_remove() and reading of the list;**

Once the above is OK, I will schedule a re-work and commit in the next office day.

rlubos commented:
The proposal looks very good, at least from my point of view, thanks!

pkchan commented:
Changed according to proposal (ref: [here](https://github.com/NordicPlayground/fw-nrfconnect-nrf/pull/1163#issuecomment-534020263)).

Rebased to daily master (79e049a, synchronized 4am 25th September, 2019 HKT), resolved conflicts in at_cmd.c and lte_lc.c. Both at_client and asset_tracker have been briefly tested for functionality.

NordicBuilder commented:
Found the following issues, please fix and resubmit:

## checkpatch issues

```
No additional types will be considered - file 'scripts/checkpatch/typedefsfile': No such file or directory
-:20: WARNING:NEW_TYPEDEFS: do not add new typedefs
#20: FILE: drivers/at_cmd/at_cmd.c:33:
+typedef struct {

-:44: ERROR:CODE_INDENT: code indent should use tabs where possible
#44: FILE: drivers/at_cmd/at_cmd.c:67:
+^I                                       at_cmd_handler_t handler,$

-:45: ERROR:CODE_INDENT: code indent should use tabs where possible
#45: FILE: drivers/at_cmd/at_cmd.c:68:
+^I                                       notif_handler_t **prev_out)$

-:68: WARNING:CONSTANT_COMPARISON: Comparisons should place the constant on the right side of the test
#68: FILE: drivers/at_cmd/at_cmd.c:91:
+	if (NULL != find_notif_handler(list, handler, &to_ins)) {

- total: 2 errors, 2 warnings, 168 lines checked

NOTE: For some of the reported defects, checkpatch may be able to
      mechanically convert to the typical style using --fix or --fix-inplace.

NOTE: Whitespace errors detected.
      You may wish to use scripts/cleanpatch or scripts/cleanfile

Your patch has style problems, please review.

NOTE: Ignored message types: AVOID_EXTERNS BRACES CONFIG_EXPERIMENTAL CONST_STRUCT DATE_TIME FILE_PATH_CHANGES MINMAX MULTISTATEMENT_MACRO_USE_DO_WHILE NETWORKING_BLOCK_COMMENT_STYLE PRINTK_WITHOUT_KERN_LEVEL SPLIT_STRING VOLATILE

NOTE: If any of the errors are false positives, please report
      them to the maintainers.



```
## Gitlint issues


Commit 9a35bda8de:
1: UC6 Body has no content, should at least have 1 line.

rlubos commented on drivers/at_cmd/at_cmd.c:
```diff
@@ -51,6 +59,93 @@ struct callback_work_item {
+
+	k_mutex_lock(&list_mtx, K_FOREVER);
+
+	SYS_SLIST_FOR_EACH_CONTAINER(&notif_handler_list, curr, node) {
```
We've agreed to use `SYS_SLIST_FOR_EACH_CONTAINER_SAFE` here.

pkchan replied:
My bad. I will fix in the next upload.

rlubos commented on drivers/at_cmd/at_cmd.c:
```diff
@@ -289,12 +384,12 @@ int at_cmd_write(const char *const cmd,
 
 void at_cmd_set_notification_handler(at_cmd_handler_t handler)
 {
+	(void)append_notif_handler(&notif_handler_list, handler);
```
I think that we should break the old API and return an error code in case of failure. All in all this function can fail now.

pkchan replied:
Should the modules calling this API be modified in this PR, too?
They include lte_lc.c, at_host.c and modem_info.c.

rlubos replied:
If you feel you have time for that, then go ahead, I don't push for it though, it can be done later on.

pkchan replied:
I would propose doing it later.

pkchan commented:
Updated:
- From SYS_SLIST_FOR_EACH_CONTAINER() to SYS_SLIST_FOR_EACH_CONTAINER_SAFE();
- Removed extra-space in notification_handler().

lemrey commented on drivers/at_cmd/at_cmd.c:
```diff
@@ -26,13 +27,20 @@ static K_THREAD_STACK_DEFINE(socket_thread_stack, \
+static K_MUTEX_DEFINE(list_mtx);
+
+/**@brief Link list element for notification handler. */
+typedef struct {
```
Avoid `typedefs` if possible. I reckon `checkpatch` gives a warning about that.

pkchan commented:
Removed typedef as suggested by @lemrey  (ref: [here](https://github.com/NordicPlayground/fw-nrfconnect-nrf/pull/1163#discussion_r328975681)).

Rebased to daily master (a9bd003, as of 4 am 27th September, 2019 HKT).

evenl commented on drivers/at_cmd/at_cmd.c:
```diff
@@ -51,6 +59,94 @@ struct callback_work_item {
+	}
+
+	/* Allocate memory and fill. */
+	to_ins = (struct notif_handler *)k_malloc(sizeof(struct notif_handler));
```
@rlubos Doesn't this going to give us the same headache as before?

rlubos replied:
`struct notif_handler` is a small structure - 8 bytes, previously we've dealt with over 2,5K byte buffers. It will not hog the heap even with multiple allocations. Perhaps a preallocated mem slab for handlers would be a better approach, I don't find it to be a blocker.

evenl commented on drivers/at_cmd/at_cmd.c:
```diff
@@ -26,13 +27,20 @@ static K_THREAD_STACK_DEFINE(socket_thread_stack, \
 				CONFIG_AT_CMD_THREAD_STACK_SIZE);
 
 static K_SEM_DEFINE(cmd_pending, 1, 1);
+static K_MUTEX_DEFINE(list_mtx);
```
Why do you use separate mutex for the notification handler slist? Doesn't that mean that the application can end up modifying the slist when it shouldn't?

evenl replied:
Ok, I see that this has be discussed before

evenl commented on drivers/at_cmd/at_cmd.c:
```diff
@@ -324,6 +420,7 @@ static int at_cmd_driver_init(struct device *dev)
 
 int at_cmd_init(void)
 {
+	sys_slist_init(&notif_handler_list);
```
What will happen if this driver is automatically initialized? The the notif_handler_list will not be initialized?

rlubos replied:
That's a good point, @pkchan this needs to be fixed.

evenl commented:
There is one problem as I see it that is not addressed by this PR. Because one socket is shared by multiple threads through this driver, also the state of that socket is shared. That is, for instance the CEREG state, the SMS reception state, or the state of any notification type. 

In the current implementation could for instance one thread enable some notification, and a separate thread disable it, the first one is still expecting to get notifications and has no knowledge that it was disabled. Another example is who decides which CEREG level should be set, if one thread request 1 and another requests level 2.

So either we must pray that all threads cooperate nicely (in a perfect world) or this is the scope of a separate service that don't belong in this driver. I opt for the second option, and that is why I didn't create this code in the first place.

stig-bjorlykke commented:
> There is one problem as I see it that is not addressed by this PR. Because one socket is shared by multiple threads through this driver, also the state of that socket is shared. That is, for instance the CEREG state, the SMS reception state, or the state of any notification type.

I also have this concerns. In the described scenario one thread may cancel notifications for another thread, and this will give unexpected results.

I think this PR should be put on hold until this issue is solved.

rlubos commented:
@evenl @stig-bjorlykke The truth is that the current approach is even worse - one module (let's say modem_info) can overwrite the notification handler of another module (let's say lwm2m_carrier) - in that case lwm2m_carrier will receive no notifications at all. I fail to see how don't we consider this PR as an improvement.

lemrey commented:
This is indeed an improvement of `at_cmd` implementation that solves a known limitation. @stig-bjorlykke  and @evenl concerns are valid but I don't see how this change precludes finding a solution to the problem that they describe, which is different from the one addressed by this PR.

rlubos commented:
@pkchan I had a chat with IoT team (@evenl and @stig-bjorlykke) about their concerns. The point is they had a slightly different idea of the design of the AT notification dispatcher.

While we all have no doubt that this PR is an improvement for the AT notification handling, their idea was to create a separate module on top of the `at_cmd` driver. As this module will need to be extended in the future with some extra logic to notification dispatching (as pointed out by @jtguggedal  and @evenl), having it separate might prevent future bloat of the `at_cmd` driver.

In that case, I might need to ask you for a small rework of this PR:
1. Create a new module for the AT notification dispatcher, I think it should lie within `lib/`, next to the parser (I'm no good at names, but perhaps `at_ntf_dispatcher.c/h` would be a good candidate),
2. Move the list and registration/deregistration API into that module,
3. Switch current users of the `at_cmd_set_notification_handler()` to the new API,
4. In this scenario, `at_cmd.c/h` would remain intact. The new module would register the notification with `at_cmd_set_notification_handler()`.

In the future, the new module will be extended with additional logic (likely by the IoT team). @evenl @stig-bjorlykke Please comment if the above description matches your idea.

pkchan commented:
Guys, I do not see any need to add extra module.

@evenl mentioned 2 concerns (ref: [here](https://github.com/NordicPlayground/fw-nrfconnect-nrf/pull/1163#issuecomment-537002455)):
1. State of the socket shared by multiple software modules
2. Enabling and disabling of events by different software modules

Here are my comments:

-1. State of the shared socket
There seems to be a wish to have a dedicated AT-command socket which has different state and event reporting for each software module. The 3GPP standard does NOT assume so. 3GPP TS27.007 Figure 1 depicts that there is only one AT-command channel for ALL applications. Before implementation, please consider contributing to TS27.007 so that it defines the behavior of multiple AT-commands.

For event levels, you can simply maximize the event level in lte_lc. Events of lower levels have parameters which are subsets of higher level events. For example, +CEREG, if lte_lc subscribed level 5, your module (e.g. modem_info) can still use level 2 parser to deal with the level 5 events.

Besides, 3GPP TS27.007 ch. 4.0 states the way how unknown parameters shall be handled:
“TE software implementors must take into account that future versions of this specification may include additional parameters beyond what is expected in any (final or intermediate) response to an AT set command, read command or test command, and beyond what is expected in any unsolicited result code. **Implementations must therefore analyse all parameters provided from the TA and discard (ignore) any parameters received following the parameters expected by the TE software.**”

-2. Enabling and disabling of events
For enabling / disabling of events, your implementation simply ignore events when they are not in use. This is somehow similar to how softdevice events are handled in regular SDK. When they are not used, they are just skipped. It should be OK here in nRF Connect SDK, too.

lemrey commented:
Well, that makes sense to me. As long as AT command responses are broadcast, I see no problem in letting each application decide what to do with it, whether to ignore it or parse it.

NordicBuilder commented:
**Some checks failed. Please fix and resubmit.**

## checkpatch issues

```
No additional types will be considered - file 'scripts/checkpatch/typedefsfile': No such file or directory
-:391: ERROR:TRAILING_WHITESPACE: trailing whitespace
#391: FILE: lib/at_notif/at_notif.h:35:
+ * @param context      Pointer to context provided by the module which has $

- total: 1 errors, 0 warnings, 432 lines checked

NOTE: For some of the reported defects, checkpatch may be able to
      mechanically convert to the typical style using --fix or --fix-inplace.

NOTE: Whitespace errors detected.
      You may wish to use scripts/cleanpatch or scripts/cleanfile

Your patch has style problems, please review.

NOTE: Ignored message types: AVOID_EXTERNS BRACES CONFIG_EXPERIMENTAL CONST_STRUCT DATE_TIME FILE_PATH_CHANGES MINMAX MULTISTATEMENT_MACRO_USE_DO_WHILE NETWORKING_BLOCK_COMMENT_STYLE PRINTK_WITHOUT_KERN_LEVEL SPLIT_STRING VOLATILE

NOTE: If any of the errors are false positives, please report
      them to the maintainers.
```

## Codeowners issues

```
New files added that are not covered in CODEOWNERS:

lib/at_notif/CMakeLists.txt
lib/at_notif/at_notif.c
lib/at_notif/at_notif.h
lib/at_notif/at_notif.rst

Please add one or more entries in the CODEOWNERS file to cover those files
```



*Tip: The bot edits this comment instead of posting a new one, so you can check the comment's history to see earlier messages.*

pkchan commented:
Per @rlubos 's request,

> In that case, I might need to ask you for a small rework of this PR:
> 
> 1. Create a new module for the AT notification dispatcher, I think it should lie within `lib/`, next to the parser (I'm no good at names, but perhaps `at_ntf_dispatcher.c/h` would be a good candidate),
> 2. Move the list and registration/deregistration API into that module,
> 3. Switch current users of the `at_cmd_set_notification_handler()` to the new API,
> 4. In this scenario, `at_cmd.c/h` would remain intact. The new module would register the notification with `at_cmd_set_notification_handler()`.

- Created module at_notif, which is initialized before at_host (controlled by Kconfig/CMakeList sequence).
- Added registration and de-registration APIs, which are at_notif_register_handler() and at_notif_deregister_handler() respectively.
- There are three modules which depends on AT-command notifications: lte_lc, at_host and modem_info. They have been changed accordingly.
- at_cmd is intact, but I have found that at_cmd_driver_init() happens after at_host_init() and the intialization of at_notif. Functionality is still OK.

The branch has been rebased to daily master, timed 4am 8th October, 2019 HKT.

Functionality have been tested using asset_tracker with MODEM_INFO enabled, running on nRF91-DK. The callback functions lte_lc, at_host and modem_info have been verified to happen correctly.

rlubos commented:
Thank you, @evenl @stig-bjorlykke please revisit.

pkchan commented:
> Please move the public header file under `/include/`, e.g.
> 
> * `/include/at/notif.h/` or
> * `/include/at_notif.h`
> 
> Keep the `.rst` file next to the header.

Moved per request.
Rebased onto c9c833b, the daily master on 4am, 9th October, 2019 HKT.

rlubos commented:
@evenl @stig-bjorlykke Ping, can you check if you're fine with current implementation?

rlubos commented:
@evenl @stig-bjorlykke Guys really, it's been two weeks already, I just need an ACK or NACK for this PR as it was reworked per your request.

evenl commented:
Sorry about the long delay. 

One comment, the call to at_cmd_set_notification_handler() should have a protection mechanism to avoid any other threads to change the handler. Like an owner thread. I suggest storing the caller thread handle when at_cmd_set_notification_handler() changes the handler from NULL, only this thread should be allowed to change the handler or free the handler by setting it back to NULL. It is far from foolproof, but will help protecting against unintentional changes.

pkchan commented:
> Sorry about the long delay.
> 
> One comment, the call to at_cmd_set_notification_handler() should have a protection mechanism to avoid any other threads to change the handler. Like an owner thread. I suggest storing the caller thread handle when at_cmd_set_notification_handler() changes the handler from NULL, only this thread should be allowed to change the handler or free the handler by setting it back to NULL. It is far from foolproof, but will help protecting against unintentional changes.

In the last known push, at_cmd_set_notification_handler() is called only by at_notif.c, function module_init(). Only one thread is involved. Anyway, at_cmd_set_notification_handler() already has protection by mutex.

Modules from multiple threads (which can be lte_lc, at_host and modem_info) can only interface with at_notif_register_handler() & at_notif_deregister_handler(). Both functions are protected with mutex inside.

LWM2M also has reference to at_cmd_set_notification_handler(). However, it is in library, so I cannot change.

carlescufi commented:
@pkchan can you please rebase this and we will merge it unless @evenl or @stig-bjorlykke have any further objections.

pkchan commented:
> @pkchan can you please rebase this and we will merge it unless @evenl or @stig-bjorlykke have any further objections.

Rebased onto daily master (6cd13b7, 4am 23rd October, 2019 HKT). Tested using asset_tracker over nRF Cloud, so that the referenced modules at_cmd, at_notif, at_host, lte_lc and modem_info have their functions verified.

ru-fu commented on include/at_notif.rst:
```diff
@@ -0,0 +1,19 @@
+.. _at_notif_readme:
+
+AT command notification manager
```
```suggestion
AT command notifications
```

ru-fu commented on include/at_notif.rst:
```diff
@@ -0,0 +1,19 @@
+AT command notification manager
+###############################
+
+AT-command notifications can be dispatched to registered modules. Modules can
```
```suggestion
AT command notifications can be dispatched to registered modules. 
Modules can register a callback function to receive AT command notifications as raw string.
```

ru-fu replied:
Please start every sentence on a new line.

ru-fu commented on include/at_notif.rst:
```diff
@@ -0,0 +1,19 @@
+AT-command notifications can be dispatched to registered modules. Modules can
+register a callback function to receive AT-command notifications as raw string.
+Multiple instances, which can be identified by pointers to contexts, are also
+supported. Modules can also de-register the callback function as well.
```
```suggestion
Modules can de-register the callback function to stop receiving notifications.
```

ru-fu commented on include/at_notif.rst:
```diff
@@ -0,0 +1,19 @@
+API documentation
+*****************
+
+| Header file: :file:`at_notif.h`
```
```suggestion
| Header file: :file:`include/at_notif.h`
```

ru-fu commented on include/at_notif.rst:
```diff
@@ -0,0 +1,19 @@
+*****************
+
+| Header file: :file:`at_notif.h`
+| Source file: :file:`at_notif.c`
```
```suggestion
| Source file: :file:`lib/at_notif/at_notif.c`
```

evenl commented:
> In the last known push, at_cmd_set_notification_handler() is called only by at_notif.c, function module_init(). Only one thread is involved. Anyway, at_cmd_set_notification_handler() already has protection by mutex.
> Modules from multiple threads (which can be lte_lc, at_host and modem_info) can only interface with at_notif_register_handler() & at_notif_deregister_handler(). Both functions are protected with mutex inside.

I probably did phrase my comment badly. So did not mean concurrency issues, what I means is that any other thread can at any point call at_cmd_set_notification_handler and point it to it's own handler. Then the dispatcher will not work. Maybe this is only a protection against stupid users, and I am not sure where NCS stands on that. It's far from bulletproof.

pkchan commented:
> > In the last known push, at_cmd_set_notification_handler() is called only by at_notif.c, function module_init(). Only one thread is involved. Anyway, at_cmd_set_notification_handler() already has protection by mutex.
> > Modules from multiple threads (which can be lte_lc, at_host and modem_info) can only interface with at_notif_register_handler() & at_notif_deregister_handler(). Both functions are protected with mutex inside.
> 
> I probably did phrase my comment badly. So did not mean concurrency issues, what I means is that any other thread can at any point call at_cmd_set_notification_handler and point it to it's own handler. Then the dispatcher will not work. Maybe this is only a protection against stupid users, and I am not sure where NCS stands on that. It's far from bulletproof.

@evenl , how about making it "run-once only"?
That is, in at_cmd_set_notification_handler(), if the notification_handler is not NULL, then exit the function as if nothing has happened.

ru-fu commented on include/at_notif.rst:
```diff
@@ -0,0 +1,21 @@
+raw string.  Multiple instances, which can be identified by pointers to
+contexts, are also Modules can de-register the callback function to stop
+receiving notifications.
+
```
Sorry, looks like this was messed up by the suggestions. :(

Should be like this:

AT command notifications can be dispatched to registered modules. 
Modules can register a callback function to receive AT command notifications as raw string.  
Multiple instances, which can be identified by pointers to contexts, are also supported.
Modules can de-register the callback function to stop receiving notifications.

pkchan replied:
Replaced by the paragraph above to the at_notif.rst in the last push.

evenl commented:
Maybe that is a better option. I can not comme up with a use-case scenario where you want to be able to release this to some other thread runtime....Maybe then it could be a build time option instead. Some macro to define the handler, or a weekly linked function.
